CORE-38: Merge dev to master for git reorganization.

pyrocto opened the pull request:
Bringing master up to date with current development.

Commit:
Create dev branch

Commit:
Add key/value store

kirkwood committed:
Initial commit work.

kirkwood committed:
Generalize.

kirkwood committed:
Fresh start.

kirkwood committed:
Merge branch 'add-comm' into dev

kirkwood committed:
Merge pull request #8 from kirkwood/dev-comm

Clean out comm, based on dev

kirkwood committed:
Add unit tests for Kademlia table.

tymm committed:
Sync roscala from former repository

kirkwood committed:
Updated for Alex's comments.

kirkwood committed:
Merge pull request #9 from kirkwood/kademlia-tests

Add unit tests for Kademlia table.

tymm committed:
Merge pull request #10 from rchain/dev-tymm-sync-roscala-repos

Sync roscala from former repository which includes the following changes:
https://github.com/rchain/Rosette/pull/32
https://github.com/rchain/Rosette/pull/33
https://github.com/rchain/Rosette/pull/34

Commit:
Add example K-framework files

Commit:
Merge pull request #11 from rchain/k

Add example K-framework files

DoSOfRedRiver committed:
Add script to generate PR

tymm committed:
Execute opcodes and return exit code (#14)

tymm committed:
JIRA# ROS-194: Add VirtualMachine.unwindAndApplyPrim(...) (#12)

kirkwood commented:
### Note on comms subsystem in this pull request
@leithaus 
The diff is hard to wrap one's head around, but this PR simply cleans out _everything_ that was in the `comm/` subdirectory in `master` and replaces it with the bare minimum:
* The Kademlia table implementation [here](https://github.com/rchain/rchain/tree/dev/comm/src/main/scala/coop/rchain/kademlia) and its unit tests [here](https://github.com/rchain/rchain/tree/dev/comm/src/test/scala/coop/rchain/kademlia)
* The pure Java zeromq implementation [here](https://github.com/rchain/rchain/tree/dev/comm/src/main/java)
* Protocol buffer stubs for peer-to-peer routing [here](https://github.com/rchain/rchain/tree/dev/comm/src/main/protobuf/coop/rchain/comm/protocol)

Since most of the code I've produced hasn't been reviewed, I will reintroduce it to `dev` in smaller chunks as it gets reworked.

tymm committed:
JIRA# ROS-196: Add VirtualMachine.handleVirtualMachineError(...) (#16)

DoSOfRedRiver committed:
Code cleanup and simplification

DoSOfRedRiver committed:
Restructure the scripts directory

DoSOfRedRiver committed:
Merge pull request #15 from DoSOfRedRiver/dev-alex-pr-script

JIRA# ROS-170: Script to initiate PR

kirkwood committed:
Add getter functions to Kademlia table.

Added a `find` method to locate a node with a particular key in the table. This is useful for implementing protocols, not for querying by remote nodes (see `lookup` for that). Also added the `peers` method to produce a `Seq` of all nodes in the table. This is used produce targets for broadcast messages, for example.

Added a few tests to sanity-check these functions.

kirkwood committed:
Removed trailing whitespace.

kirkwood committed:
scalafmt pass

tymm committed:
Add VirtualMachine.handleException(...) (#17)

kirkwood committed:
Merge pull request #19 from kirkwood/kademlia-node-getters

Add getter functions to Kademlia table.

kirkwood committed:
Add low-level unicast (UDP) transport. (#22)

* Ignore emacs clutter.

CORE-29 CORE-43 #comment Does the JIRA business have to be in the commit not the PR?

* Add low-level unicast (UDP) transport.

This is the very lowest level of the communications layer. There will eventually be one of these for each transport, starting with UDP. There is an abstraction for sending and receiving an array of bytes and an abstraction for nodes or peers in a network.

The UDP Comm class does not handle request-response, which I've put at a higher level. There is a dead-simple framing protocol to get right-sized byte arrays from datagrams. There is not quite enough handling of I/O exceptions, yet.

* Quick case class constructor valectomy.

kirkwood committed:
Add kademlia table and PING/PONG. (#23)

* Add kademlia table and PING/PONG.

This PR builds on the `UnicastComm` object to enable overlaying a simple network via UDP. Overlaying network `net` and attaching to a remote node `remote` can be as simple as:
```
import coop.rchain.comm._
val net = UnicastNetwork(NodeIdentifier("local node".getBytes), Endpoint("localhost", 33313, 33313))
val remote = new ProtocolNode(NodeIdentifier("alien node".getBytes), Endpoint("localhost", 33314, 33314), net)
remote.ping
```
which "homes" this view of the network at udp://localhost:33313 and pings the node at udp://localhost:33314.

The `UnicastNetwork` class builds a kademlia routing table using only PING and PONG messages. It extends the `ProtocolHandler` trait with UDP transport versions of the slightly higher-than-base-level comms functions for request-response (`roundTrip`) and `broadcast` (though this PR doesn't use `broadcast`). A `UnicastNetwork` continually receives network traffic in a dedicated thread. The tightness of the receive loop is a constant 500ms currently, controlled by a socket timeout down in the `Comm` object, and it should be parameterized in the future. Messages are handled in the order received.

Since UDP is (for all intents and purposes) connectionless, the `UnicastNetwork` handler class simulates synchronous request-response by blocking in `roundTrip` on a promise of a response. These promises are stored in a `pending` map, keyed by some information in the original message. There is a timeout in `roundTrip`, distinct from the receive timeout above, that can be tailored per call. After the timeout tolls, the promise is failed. The promise succeeds when the receiving thread is able to match an incoming message with one in the pending map.

A node's Kademlia table in this version of the protocol simply includes every remote node it learns of (because that remote node has sent it a PING).

Ultimately, all messages are serialized protocol buffers on the wire. The protocol classes are insulated from this by the `ProtocolMessage` object, which is responsible for parsing arrays of bytes into Scala objects and thereby for some measure of syntactic validation of messages. This object is pretty messy right now, and it will need a good going-over when it does more than PING and PONG; I think something more elegant will fall out of it when there's more to it. If anyone has a suggestion for how to structure it at this point, I'm all ears.

CORE-29 CORE-23 #comment Higher-level Kademlia-lite protocol

* Remove spurious return.

* Update Try[Boolean] to Try[Unit]

Feedback from @tymm

* Too lazy!

Feedback on #22 from @tymm.

* Add some clarity in for comprehension.

Also, scalafmt did some violence.

* Update for PR feedback.

* Add return types
* Simplify `toPeer`

* Explicitly return Unit instead of ignoring value.

DoSOfRedRiver commented on comm/src/main/scala/coop/rchain/comm/network.scala:
```diff
@@ -0,0 +1,116 @@
+          case Failure(ex)                         => ex.printStackTrace
+        }
+      }
+  }.start
```
Do you really need the `receiver` variable? The `Thread#start` method returns `Unit`.
Also, I found another way to write this code. Scala 2.12 has SAM support, so you can write something like this:
```val task: Runnable = () => { //body }
val receiver = new Thread(task).start()```

kirkwood commented:
I can't find this comment in github.

You're right, though. I (eventually) do need to remember the thread so it
can be stopped; there's currently no way to do that. I have a future PR
with that in it.

Chris.

On Sun, Oct 22, 2017 at 2:26 PM, DoSOfRedRiver <notifications@github.com>
wrote:

> *@DoSOfRedRiver* commented on this pull request.
> ------------------------------
>
> In comm/src/main/scala/coop/rchain/comm/network.scala
> <https://github.com/rchain/rchain/pull/13#discussion_r146144860>:
>
> > +  val comm = UnicastComm(local)
> +  val table = PeerTable(local)
> +
> +  private val receiver = new Thread {
> +    override def run =
> +      while (true) {
> +        comm.recv match {
> +          case Success(res) =>
> +            for {
> +              msg <- ProtocolMessage.parse(res)
> +            } dispatch(msg)
> +          case Failure(ex: SocketTimeoutException) => ()
> +          case Failure(ex)                         => ex.printStackTrace
> +        }
> +      }
> +  }.start
>
> Do you really need the receiver variable? The Thread#start method returns
> Unit.
> Also, I found another way to write this code. Scala 2.12 has SAM support,
> so you can write something like this:
>
> val receiver = new Thread(task).start()```
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rchain/rchain/pull/13#pullrequestreview-71045513>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABX0V46OKyX1GhTIsSlCz5hFmlEFQX2Gks5su7MZgaJpZM4P3qlP>
> .
>

kirkwood committed:
Add lookup/lookup response to protocol handler. (#24)

* Add kademlia table and PING/PONG.

This PR builds on the `UnicastComm` object to enable overlaying a simple network via UDP. Overlaying network `net` and attaching to a remote node `remote` can be as simple as:
```
import coop.rchain.comm._
val net = UnicastNetwork(NodeIdentifier("local node".getBytes), Endpoint("localhost", 33313, 33313))
val remote = new ProtocolNode(NodeIdentifier("alien node".getBytes), Endpoint("localhost", 33314, 33314), net)
remote.ping
```
which "homes" this view of the network at udp://localhost:33313 and pings the node at udp://localhost:33314.

The `UnicastNetwork` class builds a kademlia routing table using only PING and PONG messages. It extends the `ProtocolHandler` trait with UDP transport versions of the slightly higher-than-base-level comms functions for request-response (`roundTrip`) and `broadcast` (though this PR doesn't use `broadcast`). A `UnicastNetwork` continually receives network traffic in a dedicated thread. The tightness of the receive loop is a constant 500ms currently, controlled by a socket timeout down in the `Comm` object, and it should be parameterized in the future. Messages are handled in the order received.

Since UDP is (for all intents and purposes) connectionless, the `UnicastNetwork` handler class simulates synchronous request-response by blocking in `roundTrip` on a promise of a response. These promises are stored in a `pending` map, keyed by some information in the original message. There is a timeout in `roundTrip`, distinct from the receive timeout above, that can be tailored per call. After the timeout tolls, the promise is failed. The promise succeeds when the receiving thread is able to match an incoming message with one in the pending map.

A node's Kademlia table in this version of the protocol simply includes every remote node it learns of (because that remote node has sent it a PING).

Ultimately, all messages are serialized protocol buffers on the wire. The protocol classes are insulated from this by the `ProtocolMessage` object, which is responsible for parsing arrays of bytes into Scala objects and thereby for some measure of syntactic validation of messages. This object is pretty messy right now, and it will need a good going-over when it does more than PING and PONG; I think something more elegant will fall out of it when there's more to it. If anyone has a suggestion for how to structure it at this point, I'm all ears.

CORE-29 CORE-23 #comment Higher-level Kademlia-lite protocol

* Remove spurious return.

* Update Try[Boolean] to Try[Unit]

Feedback from @tymm

* Too lazy!

Feedback on #22 from @tymm.

* Add some clarity in for comprehension.

Also, scalafmt did some violence.

* Add lookup and lookup response.

This completes as much of the Kademlia protocol as we need in order to create and maintain a network. Particulars:

* Moved return header of protocol buffers up to top-level
* Adjusted encode/decode return types as suggested by @tymm in #22
* scalamft'd the kademlia test cases

* Update for PR feedback.

* Add return types
* Simplify `toPeer`

* Explicitly return Unit instead of ignoring value.

* Updates for PR commentary.

* Remove `???` case that can't be reached.
* Remove catch-all-exceptions in favor of `Try` block.

pyrocto commented:
Since we decided to involve @leithaus not in every push to master but rather when we cut releases, I'm going to merge this now.
